cmake/legate_helper_functions.cmake

```diff
@@ -0,0 +1,384 @@
+  "${header_output}"
+  "void ${target}_perform_registration();"
+)
+
+set(install_info_in
+[=[
+from pathlib import Path
+
+def get_libpath():
+    import os, sys, platform
+    join = os.path.join
+    exists = os.path.exists
+    dirname = os.path.dirname
+    cn_path = dirname(dirname(__file__))
+    so_ext = {
+        "": "",
+        "Java": ".jar",
+        "Linux": ".so",
+        "Darwin": ".dylib",
+        "Windows": ".dll"
+    }[platform.system()]
+
+    def find_lib(libdir):
+        target = f"libhello{so_ext}*"
+        search_path = Path(libdir)
+        matches = [m for m in search_path.rglob(target)]
+        if matches:
+          return matches[0].parent
+        return None
+
+    return (
+	find_lib("@CMAKE_BINARY_DIR@") or
+        find_lib(join(dirname(dirname(dirname(cn_path))), "lib")) or
+        find_lib(join(dirname(dirname(sys.executable)), "lib")) or
+        ""
+    )
+
+libpath: str = get_libpath()
+header: str = """@header_content@"""
+]=])
+
+set(libpath "")
+string(CONFIGURE "${install_info_in}" install_info @ONLY)
+file(WRITE ${CMAKE_SOURCE_DIR}/${target}/install_info.py "${install_info}")
+endfunction()
+
+function(legate_default_python_install target)
+  set(options)
+  set(one_value_args EXPORT)
+  set(multi_value_args)
+  cmake_parse_arguments(
+    OPT
+    "${options}"
+    "${one_value_args}"
+    "${multi_value_args}"
+    ${ARGN}
+  )
+
+  if (NOT OPT_EXPORT)
+    message(FATAL_ERROR "Need EXPORT name for legate_default_python_install")
+  endif()
+
+  if (SKBUILD)
+    add_library(${target}_python INTERFACE)
+    add_library(legate::${target}_python ALIAS ${target}_python)
+    target_link_libraries(${target}_python INTERFACE legate::core legate::${target})
+
+    install(TARGETS ${target}_python
+            DESTINATION ${lib_dir}
+            EXPORT ${OPT_EXPORT})
+
+    legate_include_rapids()
+    rapids_export(
+      INSTALL ${target}_python
+      EXPORT_SET ${OPT_EXPORT}
+      GLOBAL_TARGETS ${target}_python
+      NAMESPACE legate::
+    )
+  endif()
+endfunction()
+
+function(legate_add_cpp_subdirectory dir target)
+  set(options)
+  set(one_value_args EXPORT)
+  set(multi_value_args)
+  cmake_parse_arguments(
+    OPT
+    "${options}"
+    "${one_value_args}"
+    "${multi_value_args}"
+    ${ARGN}
+  )
+
+  if (NOT OPT_EXPORT)
+    message(FATAL_ERROR "Need EXPORT name for legate_default_install")
+  endif()
+
+  legate_include_rapids()
+
+  rapids_find_package(legate_core CONFIG
+          GLOBAL_TARGETS legate::core
+          BUILD_EXPORT_SET ${OPT_EXPORT}
+          INSTALL_EXPORT_SET ${OPT_EXPORT})
+
+  if (SKBUILD)
+    if (NOT DEFINED ${target}_ROOT)
+      set(${target}_ROOT ${CMAKE_SOURCE_DIR}/build)
+    endif()
+    rapids_find_package(${target} CONFIG
+      GLOBAL_TARGETS legate::${target}
+      BUILD_EXPORT_SET ${OPT_EXPORT}
+      INSTALL_EXPORT_SET ${OPT_EXPORT})
+    if (NOT ${target}_FOUND)
+      add_subdirectory(${dir} ${CMAKE_SOURCE_DIR}/build)
```

Comment:
I would use something like `${CMAKE_SOURCE_DIR}/legate_${${target}}_build` instead of `${CMAKE_SOURCE_DIR}/build`.

The `build` directory could already be used by a target named that, or a user might want to call `legate_add_cpp_subdirectory` multiple time and this wouldn't work

Reply:
@robertmaynard that's actually the point of this line, since this is only executed when running via scikit-build

1. If the user configures C++ independently, they can do it into `./build`
2. If the user lets scikit-build configure the C++ lib as a side-effect of building the Python lib, the C++ libs still end up in `./build` (where normally they'd be under something like `_skbuild/<arch>/cmake-build`)

This stability helps with the runtime library loading we do when dev'ing locally w/ `pip install --editable .`

Reply:
I split the difference since it's a valid point that this only supports a single source directory. I put everything in:
`${CMAKE_SOURCE_DIR}/build/legate_${target}` - which I think is okay? @robertmaynard 

This still makes the libraries easy to find.

Reply:
Maybe we should make it configurable, or do something like `build/${cmake_build_type}`? So that way at least it's `build/Debug` and `build/Release`?

Reply:
When using `add_subdirectory(${dir} ${CMAKE_SOURCE_DIR}/build)` you are stating that CMake should parse the `CMakeLists.txt` in `${dir}` and place any build information in `${CMAKE_SOURCE_DIR}/build`. This has the following issues:

1. You are placing build artifacts in the soure tree. The path should start with `CMAKE_BUILD_DIR`.
2. The second argument must be unique across a CMake project. So `${CMAKE_BUILD_DIR}/legate_${target}` would be sufficient. So the following logic doesn't work currently:
```cmake
legate_add_cpp_subdirectory(legate_plugin1 hello EXPORT hello-export)
legate_add_cpp_subdirectory(src hello EXPORT hello-export)
```

Since `legate_add_cpp_subdirectory` is executed by the users project we have no idea what source tree layout they have and shouldn't try to write to `CMAKE_SOURCE_DIR`.

Reply:
Thanks all for the suggestions. Silly mistakes on my part. I believe the latest version correctly puts things into `CMAKE_BINARY_DIR` into unique folders.

PTAL. @robertmaynard @trxcllnt

Reply:
Updated version looks good to me